pkg/plugins/runtime/k8s/controllers/cni_taint_controller.go

```diff
@@ -0,0 +1,197 @@
+package controllers
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"golang.org/x/exp/slices"
+	kube_core "k8s.io/api/core/v1"
+	kube_apierrs "k8s.io/apimachinery/pkg/api/errors"
+	kube_types "k8s.io/apimachinery/pkg/types"
+	kube_ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	kube_handler "sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	kube_reconcile "sigs.k8s.io/controller-runtime/pkg/reconcile"
+	kube_source "sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/metadata"
+)
+
+const nodeReadinessTaintKey = "NodeReadiness"
+const nodeIndexField = "spec.nodeName"
+const parentAppLabel = "parent.app"
+
+type CniNodeTaintReconciler struct {
+	kube_client.Client
+	Log logr.Logger
+}
+
+func (r *CniNodeTaintReconciler) Reconcile(ctx context.Context, req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	log := r.Log.WithValues("node", req.NamespacedName)
+	log.V(1).Info("event received")
+
+	// Fetch the Node instance
+	node := &kube_core.Node{}
+	if err := r.Get(ctx, req.NamespacedName, node); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			log.Error(err, "node not found")
+			return kube_ctrl.Result{}, nil
+		}
+		log.Error(err, "unable to fetch node")
+		return kube_ctrl.Result{}, err
+	}
+	log.Info("node successfully fetched")
+
+	kubeSystemPods := &kube_core.PodList{}
+	namespaceOption := kube_client.InNamespace("kube-system")
+	matchingFields := kube_client.MatchingFields{nodeIndexField: node.Name}
+	matchingLabels := kube_client.MatchingLabels{parentAppLabel: metadata.KumaCNI}
```

Comment:
Maybe these filters should be in `podToNodeMapper` it would avoid a lot of unecessary reconcilations?

Reply:
Yeah, adding a filter there would be a good way to reduce the number of reconciliations.